File: Haskell/2022/09.hs
import Data.List
import Data.Ord (clamp)
import Lib
import Prelude hiding (Left, Right)

data Direction = Up | Down | Left | Right deriving (Show)

type Position = (Int, Int)

type Rope = [Position]

type State = (Rope, [Position])

type Motion = (Direction, Int)

type Input = [Motion]

main :: IO ()
main = aoc 2022 9 setup solve1 solve2 ["1", "2"]

solve1 :: Input -> Int
solve1 = solve 2

solve2 :: Input -> Int
solve2 = solve 10

solve :: Int -> Input -> Int
solve n = length . nub . snd . foldl step (initialRope n, []) . directions

step :: State -> Direction -> State
step (rope, visited) m = (newRope, head newRope : visited)
  where
    newRope = stepRope rope m

stepRope :: Rope -> Direction -> Rope
stepRope r m = scanr follow (move m $ last r) $ init r

follow :: Position -> Position -> Position
follow (x, y) (x', y')
  | abs (x - x') <= 1 && abs (y - y') <= 1 = (x, y)
  | otherwise = (x + clamp (-1, 1) (x' - x), y + clamp (-1, 1) (y' - y))

directions :: [Motion] -> [Direction]
directions = concatMap $ uncurry $ flip replicate

setup :: String -> Input
setup = map parseMotion . lines

parseMotion :: String -> Motion
parseMotion (d : ' ' : n) = (parseDirection d, read n)

parseDirection :: Char -> Direction
parseDirection 'U' = Up
parseDirection 'D' = Down
parseDirection 'L' = Left
parseDirection 'R' = Right

move :: Direction -> Position -> Position
move Up (x, y) = (x, y - 1)
move Down (x, y) = (x, y + 1)
move Left (x, y) = (x - 1, y)
move Right (x, y) = (x + 1, y)

initialRope :: Int -> Rope
initialRope = flip replicate (0, 0)

File: Python/2022/09.py
from lib import *

input = read_input(2022, 9)

lines = [(*NEIGH_DICT[line[0]], int(line[2:])) for line in input.splitlines()]


def solve(lines, n):
    knots: list[tuple[int, int]] = [(0, 0) for _ in range(n)]
    visited = {knots[-1]}
    for dx, dy, cnt in lines:
        for _ in range(cnt):
            knots[0] = knots[0][0] + dx, knots[0][1] + dy
            for i in range(1, n):
                p = knots[i - 1][0] - knots[i][0]
                q = knots[i - 1][1] - knots[i][1]
                if abs(p) <= 1 and abs(q) <= 1:
                    continue
                knots[i] = knots[i][0] + max(-1, min(p, 1)), knots[i][1] + max(-1, min(q, 1))
            visited.add(knots[-1])
    return len(visited)


print(solve(lines, 2))
print(solve(lines, 10))
